pytorch_lightning/__init__.py

```diff
@@ -26,8 +26,8 @@
     from logging import getLogger
     _logger = getLogger("lightning")
 
+    from pytorch_lightning.trainer import Trainer  # Initiaized first due to state
```

Comment:
can you explain this? it feels like requiring imports in a certain order would lead to bugs slipping into the codebase more easily

Reply:
I did a quick test and I think it is because pytorch_lightning.trainer.state.TrainerMode is included in a cyclic import.
For example, if I move TrainerMode to pytorch_lightning.overrides.data_parallel, then the import order doesn't matter (tests don't break in either case). 
I'm not saying it should go there but I would try to move it out of the import loop.

Reply:
It's due to where `state.py` is located. Any suggestions on whether I should move it out like what @awaelchli said or should I keep it as it is for now?

Reply:
hmm yeah it'd be preferable if we didn't have to rely on import order for things to work properly. do you know where the import cycle is occuring?


```
# pytorch_lightning/trainer/state.py
(TrainerMode class is defined)

# pytorch_lightning/__init__.py
from pytorch_lightning.trainer import Trainer 
from pytorch_lightning.core import LightningModule

# pytorch_lightning/overrides/data_parallel.py
from pytorch_lightning.trainer.state import TrainerMode

# pytorch_lightning/trainer/evaluation_loop.py
from pytorch_lightning.trainer.state import TrainerMode

# pytorch_lightning/trainer/trainer.py
from pytorch_lightning.trainer.state import TrainerMode
```

just looking at the imports from this PR i'm not seeing it

Reply:
Maybe it's not cyclic import. 
I noticed this:
- Flip the imports @jeremyjordan highlighted (LightningModule first, then Trainer)
- Open interactive python shell and import TrainerMode: 
`from pytorch_lightning.trainer.state import TrainerMode`
- It tries to import `LightningDistributedDataParallel`, `TrainerDataLoadingMixin` and `LightningModule`

It must be a problem with imports in `pytorch_lightning/__init__.py` or `pytorch_lightning/trainer/__init__.py`

Reply:
I think I found the cycle:
When we do
`from pytorch_lightning.trainer.state import TrainerMode` it will run the init from `pytorch_lightning`, which imports Trainer. Trainer tries to import 
`from pytorch_lightning.trainer.state import TrainerMode` and so on ...

Reply:
I'll take a look at it tomorrow and fix it so that import order is not relied on. :)

Reply:
I think it's either `TrainerMode` is moved out of `trainer` or `TrainerMode` needs to be imported first like:
```python
from pytorch_lightning.trainer.state import TrainerMode
from pytorch_lightning.core import LightningModule
from pytorch_lightning.trainer import Trainer
from pytorch_lightning.callbacks import Callback
from pytorch_lightning.core import data_loader
```
however, order still matters even in this case. the main issue lies with the import in `overrides/data_parallel.py`. any suggestions? @jeremyjordan @awaelchli @Borda

Reply:
I would stay outside and call it `states` as the trainer status will be added